core/services/mongodb-remote/src/test/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizerUnitTests.kt

```diff
@@ -1,190 +1,882 @@
+            else ChangeEvent.changeEventForLocalInsert(ctx.namespace, expectedDocument, true)
+            ctx.verifyChangeEventListenerCalledForActiveDoc(
+                times = if (shouldWaitForError) 0 else 1,
+                expectedChangeEvent = if (shouldWaitForError) null else expectedChangeEvent)
+            ctx.verifyConflictHandlerCalledForActiveDoc(times = 1)
+            ctx.verifyErrorListenerCalledForActiveDoc(times = if (shouldWaitForError) 1 else 0,
+                error = if (shouldWaitForError) ctx.exceptionToThrowDuringConflict else null)
+        }
+    }
+
+    private val harness = SyncTestHarness()
+
     @After
     fun teardown() {
+        harness.teardown()
         CoreRemoteClientFactory.close()
         ServerEmbeddedMongoClientFactory.getInstance().close()
     }
 
-    private val namespace = MongoNamespace("foo", "bar")
-    private val networkMonitor = object : NetworkMonitor {
-        override fun removeNetworkStateListener(listener: NetworkMonitor.StateListener) {
-        }
+    @Test
+    fun testNew() {
+        val ctx = harness.freshTestContext(shouldPreconfigure = false)
 
-        override fun isConnected(): Boolean {
-            return true
-        }
+        // a fresh, non-configured dataSynchronizer should not be running.
+        assertFalse(ctx.dataSynchronizer.isRunning)
+    }
 
-        override fun addNetworkStateListener(listener: NetworkMonitor.StateListener) {
-        }
+    @Test
+    fun testOnNetworkStateChanged() {
+        val ctx = harness.freshTestContext()
+
+        // verify that, since we are offline, start has not been called
+        ctx.isOnline = false
+        assertFalse(ctx.dataSynchronizer.isRunning)
+        ctx.verifyStartCalled(0)
+        ctx.verifyStopCalled(2)
+
+        // verify that, since we are online, the dataSync has started
+        ctx.isOnline = true
+        ctx.verifyStartCalled(1)
+        ctx.verifyStopCalled(2)
+    }
+
+    @Test
+    fun testStartAndStop() {
+        val ctx = harness.freshTestContext(shouldPreconfigure = false)
+        assertFalse(ctx.dataSynchronizer.isRunning)
+        ctx.reconfigure()
+
+        // with a configuration, we should be running
+        assertTrue(ctx.dataSynchronizer.isRunning)
+
+        ctx.dataSynchronizer.stop()
+        assertFalse(ctx.dataSynchronizer.isRunning)
+    }
+
+    @Test
+    fun testSuccessfulInsert() {
+        val ctx = harness.freshTestContext()
+
+        // insert the doc, wait, sync, and assert that the expected change events are emitted
+        ctx.insertTestDocument()
+        ctx.waitForEvent()
+        ctx.verifyChangeEventListenerCalledForActiveDoc(
+            times = 1,
+            expectedChangeEvent = ChangeEvent.changeEventForLocalInsert(
+                ctx.namespace,
+                ctx.testDocument,
+                true))
+        ctx.doSyncPass()
+        ctx.waitForEvent()
+        ctx.verifyChangeEventListenerCalledForActiveDoc(
+            times = 1,
```

Comment:
[overarching q] Since we're checking that the change event listener is called more than one time (once before doing the sync pass, and once after), does that mean end users will have `onEvent` called multiple times in their change event listeners for a single insert operation? that seems like incorrect behavior?

If that's not what's being tested here, can we clarify in the comments why we're waiting for the events again after doSyncPass? I see this behavior in a lot of the tests here

@edaniels would appreciate your insight on this

Reply:
Discussed offline.